src/modules/Game.class.js

```diff
@@ -0,0 +1,254 @@
+      const mergedCellsIndexes = new Set();
+      const emptyIndexes = [];
+
+      for (let i = 0; i < row.length; i++) {
+        if (row[i] === 0) {
+          emptyIndexes.push(i);
+          continue;
+        }
+
+        let cellIndex = i;
+
+        if (emptyIndexes.length) {
+          cellIndex = emptyIndexes.shift();
+          emptyIndexes.push(i);
+
+          row[cellIndex] = row[i];
+          row[i] = 0;
+          isStateChanged = true;
+        }
+
+        const previousCellIndex = cellIndex - 1;
+
+        if (
+          previousCellIndex >= 0 &&
+          row[previousCellIndex] === row[cellIndex] &&
+          !mergedCellsIndexes.has(previousCellIndex)
+        ) {
+          mergedCellsIndexes.add(previousCellIndex);
+          row[previousCellIndex] *= 2;
+          row[cellIndex] = 0;
+          emptyIndexes.unshift(cellIndex);
+          this.#score += row[previousCellIndex];
+          isStateChanged = true;
+
+          if (row[previousCellIndex] >= 2048) {
+            this.#status = 'win';
+          }
+        }
+      }
+    });
+
+    if (isStateChanged) {
+      this.#addNumber();
+    }
+
+    if (!this.#isForCheck && !this.#isMovePossible()) {
+      this.#status = 'lose';
+    }
+
+    return isStateChanged;
+  }
+  moveRight() {
+    this.#reverse();
+
+    const result = this.moveLeft();
+
+    this.#reverse();
+
+    return result;
+  }
+  moveUp() {
+    this.#rotate();
+
+    const result = this.moveLeft();
+
+    this.#rotate();
+
+    return result;
+  }
+  moveDown() {
+    this.#rotate();
+
+    const result = this.moveRight();
+
+    this.#rotate();
+
+    return result;
+  }
+
+  /**
+   * @returns {number}
+   */
+
+  getScore() {
+    return this.#score;
+  }
+
+  /**
+   * @returns {number[][]}
+   */
+
+  getState() {
+    return this.#state.map((row) => [...row]);
+  }
+
+  /**
+   * Returns the current game status.
+   *
+   * @returns {string} One of: 'idle', 'playing', 'win', 'lose'
+   *
+   * `idle` - the game has not started yet (the initial state);
+   * `playing` - the game is in progress;
+   * `win` - the game is won;
+   * `lose` - the game is lost
+   */
+  getStatus() {
+    return this.#status;
+  }
+
+  /**
+   * Starts the game.
+   */
+  start() {
+    this.#addNumber();
+    this.#addNumber();
+    this.#status = 'playing';
+  }
+
+  /**
+   * Resets the game.
+   */
+  restart() {
+    this.#applyInitialState();
+    this.#score = 0;
+    this.#status = 'idle';
+  }
+
+  #applyInitialState() {
+    this.#state = this.#initialState.map((row) => [...row]);
+  }
+  /**
+   * Checks if any move is possible
+   * @returns {boolean}
+   */
+  #isMovePossible() {
+    const gameCopy = new Game(this.getState(), true);
+
+    return (
+      gameCopy.moveLeft() ||
+      gameCopy.moveRight() ||
+      gameCopy.moveUp() ||
+      gameCopy.moveDown()
+    );
+  }
+  /**
+   * Returns if adding number to state was succesfull
+   * @returns {boolean}
+   */
+  #addNumber() {
```

Comment:
The method '#addNumber' has a side effect of adding a number to the game state. If this method is called unintentionally, it could lead to an incorrect game state. Consider adding a guard or making sure it's called only when necessary.